airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/cursor/default_file_based_cursor.py

```diff
@@ -0,0 +1,120 @@
+                )
+
+    def get_state(self) -> StreamState:
+        state = {
+            "history": self._file_to_datetime_history,
+        }
+        return state
+
+    def _is_history_full(self) -> bool:
+        """
+        Returns true if the state's history is full, meaning new entries will start to replace old entries.
+        """
+        return len(self._file_to_datetime_history) >= self._max_history_size
+
+    def _should_sync_file(self, file: RemoteFile, logger: logging.Logger) -> bool:
+        if file.uri in self._file_to_datetime_history:
+            # If the file's uri is in the history, we should sync the file if it has been modified since it was synced
+            updated_at_from_history = datetime.strptime(self._file_to_datetime_history.get(file.uri), self.DATE_TIME_FORMAT)
+            if file.last_modified < updated_at_from_history:
+                logger.warning(
+                    f"The file {file.uri}'s last modified date is older than the last time it was synced. This is unexpected. Skipping the file."
+                )
+            else:
+                return file.last_modified > updated_at_from_history
+            return file.last_modified > updated_at_from_history
+        if self._is_history_full():
+            if file.last_modified > self._initial_earliest_file_in_history.last_modified:
+                # If the history is partial and the file's datetime is strictly greater than the earliest file in the history,
+                # we should sync it
+                return True
+            elif file.last_modified == self._initial_earliest_file_in_history.last_modified:
+                # If the history is partial and the file's datetime is equal to the earliest file in the history,
+                # we should sync it if its uri is strictly greater than the earliest file in the history
+                return file.uri > self._initial_earliest_file_in_history.uri
+            else:
+                # Otherwise, only sync the file if it has been modified since the start of the time window
+                return file.last_modified >= self.get_start_time()
+        else:
+            # The file is not in the history and the history is complete. We know we need to sync the file
+            return True
+
+    def get_files_to_sync(self, all_files: Iterable[RemoteFile], logger: logging.Logger) -> Iterable[RemoteFile]:
+        if self._is_history_full():
+            logger.warning(
+                f"The state history is full. "
+                f"This sync and future syncs won't be able to use the history to filter out duplicate files. "
+                f"It will instead use the time window of {self._time_window_if_history_is_full} to filter out files."
+            )
+        for f in all_files:
+            if self._should_sync_file(f, logger):
+                yield f
+
+    def get_start_time(self) -> datetime:
```

Comment:
Do we have an opinion as to if datetimes should be naive or not? I'm asking because the return `datetime` does not explain the extend as to which the user will be able to use the output (for example, Python can't compare a naive datetime against a aware one)